### ISSUE TITLE
Using a `torch.compile`'d function in a multiprocessing subprocess causes process hang on exit

### ISSUE THREAD

**Issue by cbcase** [pytorch/pytorch#115366]
### 🐛 Describe the bug

I am using `torch.multiprocessing` to launch subprocesses and run vanilla PyTorch code on the GPU in those processes. Whenever a process calls into a function decorated with `torch.compile`, that process is unable to exit cleanly -- if the parent process tries to `join()` on the subprocess, it hangs indefinitely. If you remove the `compile` decoration, things work normally. Here is a small reproducer:

```
import torch
import torch.multiprocessing as mp

@torch.compile
def affine(x, a, b):
    return a * x + b

def run_fn(recv_q):
    while True:
        x_recv = recv_q.get()
        if x_recv is None:
            break
        x = torch.empty_like(x_recv)
        x.copy_(x_recv)
        del x_recv
        _ = affine(x, 3.0, 2.0)

def main():
    ctx = mp.get_context("spawn")
    send_q = ctx.SimpleQueue()
    proc = ctx.Process(
        target=run_fn,
        args=(send_q,),
        daemon=True,
    )
    proc.start()

    # Separate scope to not hold onto tensors
    def test():
        for _ in range(2):
            x = torch.randn(100, 100, device="cuda")
            send_q.put(x)

    test()
    send_q.put(None)
    print("Waiting on subprocess...")
    proc.join()
    print("Done.")

if __name__ == "__main__":
    main()
```
If I run this code as-is, then it will hang after `Waiting on subprocess...` until you CTRL-C. But if you remove the `torch.compile` decoration on the `affine` function, it exits cleanly.

### Versions

Collecting environment information...
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.4
Libc version: glibc-2.31

Python version: 3.9.18 (main, Sep 11 2023, 13:41:44)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-75-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000
GPU 2: NVIDIA RTX A5000
GPU 3: NVIDIA RTX A5000

Nvidia driver version: 535.54.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          64
On-line CPU(s) list:             0-63
Thread(s) per core:              1
Core(s) per socket:              64
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           1
Model name:                      AMD EPYC 7513 32-Core Processor
Stepping:                        1
CPU MHz:                         2599.998
BogoMIPS:                        5199.99
Virtualization:                  AMD-V
L1d cache:                       4 MiB
L1i cache:                       4 MiB
L2 cache:                        32 MiB
L3 cache:                        16 MiB
NUMA node0 CPU(s):               0-63
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core invpcid_single ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid arch_capabilities

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] torch==2.1.0+cu121
[pip3] torch-tb-profiler==0.4.1
[pip3] torchaudio==2.1.0+cu121
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.1.0
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] torch                     2.1.0+cu121              pypi_0    pypi
[conda] torch-tb-profiler         0.4.1                    pypi_0    pypi
[conda] torchaudio                2.1.0+cu121              pypi_0    pypi
[conda] torchvision               0.16.0+cu121             pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

cc @VitalyFedyunin @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

**Comment by youkaichao** [comment]
The code runs well in my MacOS (I moved the device to CPU) 👀

**Comment by youkaichao** [comment]
> The code runs well in my MacOS (I moved the device to CPU) 👀

Also run the above code in Linux + CUDA, the code runs without problem.

I'm using built-from-source PyTorch version `2.2.0a0+git95aec25`.

**Comment by cbcase** [comment]
I suppose it's possible this was fixed going from 2.1 -> 2.2? We're on the release 2.1 pip wheel (as the env info should show). Any idea what the fix might have been? I will try with a nightly wheel when I get a chance

**Comment by ezyang** [comment]
Did upgrading to nightly help?

**Comment by cbcase** [comment]
Obviously a _very_ late reply, but I ended up waiting until there was a 2.2.0 pip wheel. What I have found:
- I still have the same issue with my own container build
- I can't reproduce the issue (ie, it works fine) with Nvidia's 24.01 pytorch container (which is 2.2.0)

So I'm inclined to believe there's something wonky going on with specifically my env. I'll update if / when I have more useful debugging info.
